circuits/transfer/src/withdraw_from_obfuscated.rs

```diff
@@ -26,42 +39,104 @@ pub struct WithdrawFromObfuscatedCircuit {
 
 impl WithdrawFromObfuscatedCircuit {
     pub fn new(
-        message_r: JubJubScalar,
-        message_ssk: &SecretSpendKey,
-        message: &Message,
+        input: &Message,
+        input_psk: &PublicSpendKey,
+        input_entropy: &JubJubScalar,
+        change: &Message,
+        change_entropy: JubJubScalar,
+        change_derive_key_is_public: bool,
+        change_psk: &PublicSpendKey,
         output: &Note,
         output_vk: Option<&ViewKey>,
     ) -> Result<Self, PhoenixError> {
-        let message_psk = message_ssk.public_spend_key();
+        let (input_value, input_blinder) =
+            input.decrypt(input_entropy, input_psk)?;
 
-        let message_commitment = *message.value_commitment();
-        let (message_value, message_blinder) =
-            message.decrypt(&message_r, &message_psk)?;
+        let input_commitment = *input.value_commitment();
+
+        let (change_value, change_blinder) =
+            change.decrypt(&change_entropy, change_psk)?;
+        let change_commitment = *change.value_commitment();
+
+        let identity = JubJubAffine::identity().into();
+        let change_derive_key_a = *change_psk.A();
```

Comment:
Everything seems to be copied. Why?

Reply:
dusk-plonk API currently takes everything as owned values